### PR TITLE
Add proximity-based predictor functions

### PR DESCRIPTION
**Pull request by rfl-urbaniak**
This pull request introduces a set of functions designed to compute proximity-based predictor scores for forager interactions. 

1.   `_piecewise_proximity_function`:
        Computes a piecewise proximity function to model varying effects of distance on proximity scores.
        
2.    `_proximity_predictor_contribution`:
        Calculates the proximity score contribution of an agent to points in a grid based on distance.

3.    `_proximity_predictor`:
        Computes proximity-based predictor scores for foragers within a specified interaction length. Aggregates the impact of individual agents within local windows. Normalizes scores by scaling to the maximum absolute value.

4.    `generate_proximity`:
        Generates proximity-based predictors for a group of foragers. Retrieves parameters from a `dataObject` and computes proximity predictor values using the `_proximity_predictor` function.

5.    `filter_nearest:`
        Identifies the nearest forager from a list of potential neighbors,
        computes Euclidean distances between the focal forager and other candidates at a specified time step, returns the index of the nearest forager.

### CODE REVIEW

File: collab2/foraging/toolkit/proximity.py [hunk 1]
```diff
@@ -1,93 +1,209 @@
...
-    visibility_range,
-    getting_worse=1.5,
-    optimal=4,
-    proximity_decay=1,
-    start=0,
-    end=None,
-    time_shift=0,
+def _piecewise_proximity_function(
+    distance: Union[float, np.ndarray],
+    getting_worse: float = 1.5,
```
Comment (PalkaPuri) [inline collab2/foraging/toolkit/proximity.py:13]: can we rename `getting_worse` to `repulsion_radius` so that its meaning is more clear?

Comment (PalkaPuri) [reply collab2/foraging/toolkit/proximity.py:13]: Relatedly, maybe we can call `optimal` as `optimal_distance`

File: collab2/foraging/toolkit/proximity.py [hunk 2]
```diff
@@ -1,93 +1,209 @@
...
+
+    :return: A numpy ndarray representing the computed proximity scores for each point in the grid.
+    """
+
+    distance_to_other = np.sqrt((grid["x"] - x_other) ** 2 + (grid["y"] - y_other) ** 2)
+    proximity_score = proximity_function(distance_to_other, **proximity_function_kwargs)
+    return proximity_score
+
+
+def _proximity_predictor(
```
Comment (PalkaPuri) [inline collab2/foraging/toolkit/proximity.py:96]: pls rename to `_generate_proximity_predictor` :)

File: collab2/foraging/toolkit/proximity.py [hunk 3]
```diff
@@ -1,93 +1,209 @@
...
-            distbt = forager_distances[b].iloc[t].drop(b + 1)
+    return predictor
 
-            visible_foragers = distbt[distbt <= visibility_range].index.tolist()
 
-            if len(visible_foragers) > 0:
-                for vb in range(len(visible_foragers)):
-                    o_x = foragers[visible_foragers[vb] - 1]["x"].iloc[t]
-                    o_y = foragers[visible_foragers[vb] - 1]["y"].iloc[t]
+def generate_proximity(foragers_object: dataObject, predictor_name: str):
```
Comment (PalkaPuri) [inline collab2/foraging/toolkit/proximity.py:182]: pls rename to `generate_proximity_predictor`

**Review by PalkaPuri — CHANGES_REQUESTED** [review]

File: collab2/foraging/toolkit/filtering.py [hunk 4]
```diff
@@ -42,3 +62,46 @@ def filter_by_distance(
             foragers_ind, f, t, foragersDF, interaction_constraint_params
         )
     return foragers_ind
+
+
+def filter_nearest(
```
[the remark this answers is not in the document]
Comment (rfl-urbaniak) [reply collab2/foraging/toolkit/filtering.py:67]: sure

File: collab2/foraging/toolkit/proximity.py (hunk 1 above)
Comment (rfl-urbaniak) [reply collab2/foraging/toolkit/proximity.py:13]: sure

File: collab2/foraging/toolkit/proximity.py (hunk 2 above)
Comment (rfl-urbaniak) [reply collab2/foraging/toolkit/proximity.py:96]: done

File: collab2/foraging/toolkit/proximity.py (hunk 3 above)
Comment (rfl-urbaniak) [reply collab2/foraging/toolkit/proximity.py:182]: done

**Review by PalkaPuri — APPROVED** [review]

**Pull request merged**